pkg/kubelet/certificate/kubelet.go

```diff
@@ -92,6 +105,43 @@ func NewKubeletServerCertificateManager(kubeClient clientset.Interface, kubeCfg
 	return m, nil
 }
 
+func addressesToHostnamesAndIP(addresses []v1.NodeAddress) (dnsNames []string, ips []net.IP) {
+	seenDNS := map[string]bool{}
+	seenIP := map[string]bool{}
+	for _, address := range addresses {
+		if len(address.Address) == 0 {
+			continue
+		}
+
+		switch address.Type {
+		case v1.NodeHostName:
+			if ip := net.ParseIP(address.Address); ip != nil {
+				if !seenIP[address.Address] {
+					ips = append(ips, ip)
+					seenIP[address.Address] = true
+				}
+			} else {
+				if !seenDNS[address.Address] {
+					dnsNames = append(dnsNames, address.Address)
+					seenDNS[address.Address] = true
+				}
+			}
+		case v1.NodeExternalIP, v1.NodeInternalIP:
+			if ip := net.ParseIP(address.Address); ip != nil && !seenIP[address.Address] {
+				ips = append(ips, ip)
+				seenIP[address.Address] = true
+			}
+		case v1.NodeExternalDNS, v1.NodeInternalDNS:
+			if !seenDNS[address.Address] {
+				dnsNames = append(dnsNames, address.Address)
+				seenDNS[address.Address] = true
+			}
+		}
+	}
+
```

Comment:
@liggitt since we are using reflect to verify a deep equal later in the code, we should sort the dns and IP addresses just in case they get returned in a different order at some point.

Reply:
good point, done